lib/clients/AzureClient.py

```diff
@@ -0,0 +1,459 @@
+            next_lun = 0
+            for disk in all_data_disks:
+                if disk.lun == next_lun:
+                    next_lun += 1
+
+            existing_devices_path = glob.glob(
+                '/sys/bus/scsi/devices/*:*:*:{}/block'.format(next_lun))
+            virtual_machine.storage_profile.data_disks.append({
+                'lun': next_lun,
+                'name': volume.name,
+                'create_option': DiskCreateOptionTypes.attach,
+                'managed_disk': {
+                    'id': volume.id
+                }
+            })
+
+            disk_attach_operation = self.compute_client.virtual_machines.create_or_update(
+                self.resource_group,
+                virtual_machine.name,
+                virtual_machine
+            )
+
+            self._wait('Waiting for attachment of volume {} to get ready...'.format(volume_id),
+                       lambda operation: operation.done() is True,
+                       None,
+                       disk_attach_operation)
+
+            updated_vm = disk_attach_operation.result()
+            # TODO: Need to take care how device is handled
+            all_devices_path = glob.glob(
+                '/sys/bus/scsi/devices/*:*:*:{}/block'.format(next_lun))
+            new_devices_path = list(set(all_devices_path) -
+                                    set(existing_devices_path))
+            if len(new_devices_path) > 1:
+                raise Exception(
+                    'Found more than one new devices while attaching volume!')
+            device = '/dev/{}'.format(self.shell(
+                'ls {}'.format(new_devices_path[0])).rstrip())
+            self._add_volume_device(volume_id, device)
+            attachment = Attachment(0, volume_id, instance_id)
+            self._add_attachment(volume_id, instance_id)
+            self.logger.info(
+                '{} SUCCESS: volume-id={}, instance-id={}\n Updated vm:{}'.format(
+                    log_prefix, volume_id, instance_id, updated_vm))
+        except Exception as error:
+            message = '{} ERROR: volume-id={}, instance-id={}\n{}'.format(
+                log_prefix, volume_id, instance_id, error)
+            self.logger.error(message)
+
+            # The following lines are a workaround in case of inconsistency:
+            # The attachment process may end with throwing an Exception, e.g.
+            # 'list index out of range', but the attachment has been successful. Therefore, we must
+            # check whether the volume is attached and if yes, trigger the detachment
+            # TODO : Following need to take care: volume.status is not in-use in case of Azure
+            volume = self.compute_client.disks.get(
+                self.resource_group, volume_id)
+            if volume.managed_by is not None:
+                self.logger.warning('[VOLUME] [DELETE] Volume is attached although the attaching process failed, '
```

Comment:
why `[VOLUME] [DELETE] ` here ?

Reply:
# The following lines are a workaround in case of inconsistency:
            # The attachment process may end with throwing an Exception, e.g.
            # 'list index out of range', but the attachment has been successful. Therefore, we must
            # check whether the volume is attached and if yes, trigger the detachment

Reply:
I think the question here is why not use the log_prefix like in other cases.

Reply:
Okay... Here the function is about attachment and log_prefix is defined as `log_prefix = '[ATTACHMENT] [CREATE]'`

But my operation was delete volume  and it is once .. That's why didn't put the in a constant.